### PR TITLE
Added new SysMonitorOshi v0 using Oshi library

### LINKED ISSUE #14157: Community roadmap 2023
It would be cool to have a publicly published roadmap of things the community is working on, so I thought we should do one for 2023. The year isn't even half over, so it seems like a good time 🙂.

I'm starting this list off with some things that are on my radar. I hope everyone reading this sees some exciting features on the list! Please feel free to post here if you've got any feedback, questions, or comments about items that are or aren't here.

Btw, note that items being on this list doesn't mean they will definitely ship in 2023. It is rather a statement that there is interest in working on them, and some likelihood they will get done. So, it's a good way to get a sense for what kinds of things are receiving active development interest and activity.

### Query language

- Window functions (#13458, #13514, #13902, #13773)
- Distributed shuffling JOIN (#13506)
- Array columns (#13803)
- Array UNNEST (#13268)
- Standardize on SQL-compatible null handling (#14154)
- ORDER BY non-`__time` in native Scan query

### Query engine

- Detect unavailable segments (#13967)
- Larger, self-adjusting subquery limit (#13952)
- High-cardinality GROUP BY performance enhancements
- Interactive API for multi-stage query
- Export multi-stage query task results to object storage
- Query directly from object storage
- Update to Calcite 1.34 or later (#13532)
- EXPLAIN ANALYZE

### Data management

- Batch DELETE DML (#14145)
- Batch UPDATE, UPSERT DML
- DDL (CREATE / ALTER / DROP TABLE)
- Schema auto-discovery
- Compaction concurrent with streaming ingestion
- Metadata catalog (#12546)
- Validation of load rules, partitioning, etc
- Broadcast tables (alternative to lookups)
- Reset to arbitrary Kafka offsets

### Operations

- Java 17 full support (#12838)
- Java 21 experimental support
- Overlord scalability to larger numbers of tasks
- Coordinator scalability to larger numbers of segments
- K8s task runner (#13156 and follow-ups)
- SysMonitor on ARM

### For Druid developers

- Standardized extension API
- Improve build pipeline and testing system

### PR DESCRIPTION
**Pull request by hardikbajaj**
Fixes one of the issues in #14157. (SysMonitor on ARM in roadmap)

### Description
SysMonitor uses [Sigar](https://github.com/hyperic/sigar) library, which not only is not maintained and doesn't run on ARM64 arch. A JNA based [oshi](https://github.com/oshi/oshi) library was proposed by @abhishekagarwal87  in slack community which is well maintained, used by projects like Apache flink, and also works well on ARM architecture too. Here I introduce v0 of `SysMonitorOshi`, oshi based SysMonitor for Druid. 
This monitor monitors most of the metrics emitted by `SysMonitor.java` but is a v0 implementation which just consumes metrics emitted by Oshi directly, without modifying/mapping it.

#### Fixed the bug ...
SysMonitor not usable in ARM architecture

#### Release note

Added a new monitor `SysMonitorOshi` to replace `SysMonitor`. The new monitor has a wider support for different machine architectures including ARM instances. Please switch to SysMonitorOshi as `SysMonitor` is now deprecated and will be removed in future releases. 

<hr>

##### Key changed/added classes in this PR
 * `SysMonitorOshi.java`
 * `NoopSysMonitorOshi.java`
 * `SysMonitorOshiTest.java`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java:173]: I think that volume name should be used for fsDevName and Name for the dirName dimension. thoughts?

File: processing/src/test/java/org/apache/druid/java/util/metrics/SysMonitorOshiTest.java
Comment (abhishekagarwal87) [inline processing/src/test/java/org/apache/druid/java/util/metrics/SysMonitorOshiTest.java:32]: We should add more tests.

**Review by abhishekagarwal87 — COMMENTED** [review]
This is good. I am not fully done yet but there is some feedback that can be addressed right away.

**Comment by clintropolis** [comment]
the 'check-licenses' script we have to help us ensure that we collect all the correct licensing information for our source distributions to meet apache requirements [has entries for the licenses it is aware of to normalize them into standard buckets](https://github.com/apache/druid/blob/master/distribution/bin/check-licenses.py#L294), it looks like maybe something is reporting a string that it hasn't seen before, which looks like what this error is perhaps:
```
KeyError: 'SPDX-License-Identifier: MIT'
```
I think if you add that entry into the script and bucket it into the other MIT licenses then it should probably be ok?

Also if there are any transitive dependencies pulled in by this new library they might also need added to the licenses.yaml file. I know this stuff is somewhat tedious, so apologies for that, and still not a perfect system by any means because it is prone to becoming stale, but it does significantly reduce the burden of doing releases.

File: processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java:47]: Sure I'll do that

File: processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java:112]: Yeah, that was my concern. If someone uses SysMonitor till Oshi monitor is fully documented and tested. 
I thought we'll change back to Sys once SysMonitor is deprecated.

File: processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java:173]: I think `fs.getName()` returns volume name only. We're emitting 160 metrics in OshiSysMonitor, Let me share them on slack and you can verify if there's some issue. That would be easier

File: processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java
Comment (abhishekagarwal87) [reply processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java:112]: this will double the volume of metrics and there are tools out there already tracking these metrics. If we change the name, then those tools need to adjust. That will be painful. so let's retain the same names.

**Comment by abhishekagarwal87** [comment]
I also unlinked the community roadmap issue from this PR so that the issue doesn't get closed after this PR is merged :)

File: processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java
Comment (abhishekagarwal87) [reply processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java:402]: There is also a difference in how hardware address is being represented. E.g. 
metric, Oshi value, sigar value
```
.sys/net/read/packets.userDims.netAddress,a4:83:e7:9b:de:d4,100.217.188.78
.sys/net/read/packets.userDims.netHwaddr,[2405:201:5806:a040:cc6b:688e:7502:3991, 2405:201:5806:a040:1490:85f8:f64f:42b4, fe80:0:0:0:1406:7632:fc08:2633],00:00:00:00:00:00
```

Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java:402]: >We can leave v6 metrics out for now.

Should I add tcpv4 and tcpv6 to give tcp metrics? 

>  We also have these metrics missing in Oshi

Yeah these metrics are missing in Oshi implementation. I'll share you output of every metrics that Oshi can release

File: processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java:112]: Sounds correct. Retained

File: processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/SysMonitorOshi.java]: @abhishekagarwal87 Marking it unresolved until we figure out best way to get fsTypeName.

File: processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java:242]: Acknowledged, thanks for pointing this out. I missed it

File: processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java:268]: yes, it's supposed to return a list of Ipv4 addresses. Returning an empty array in some cases

File: processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java:326]: yes, confirmed from both Oshi SystemInfo test file and how it is implemented in SysMonitor

File: processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java:200]: doing total - usable. Thanks for pointing this out

File: processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java:67]: fsTypeWhitelist right ? Yeah you're right, now with local SysTYpe filter, we would not be needing this anymore.

File: processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java:293]: Discussed and implemented

File: processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java:221]: Going ahead with Disk metrics being emitted and left a comment if partition level metrics needs to be emitted in future

File: processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java:316]: looks like the tests were indeed helpful :)

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by abhishekagarwal87** [comment]
Approved and will merge once CI passes. Thank you @hardikbajaj for your contribution.

File: processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java
Comment (hardikbajaj) [reply processing/src/main/java/org/apache/druid/java/util/metrics/OshiSysMonitor.java:316]: Yes they were 😅

**Comment by hardikbajaj** [comment]
Thanks @abhishekagarwal87  for giving me this great learning opportunity, in form of a very meaningful Good First issue :). Also thanks for being super supportive and helping me out with your reviews, and teachings.

**Comment by abhishekagarwal87** [comment]
Merged. Appreciate the persistency @hardikbajaj :)

**Pull request merged**
